synapse/rest/client/v2_alpha/devices.py

```diff
@@ -197,21 +197,32 @@ def __init__(self, hs):
         self.device_handler = hs.get_device_handler()
 
     async def on_GET(self, request: SynapseRequest):
-        requester = await self.auth.get_user_by_req(request, allow_guest=True)
-        (
-            device_id,
-            dehydrated_device,
-        ) = await self.device_handler.get_dehydrated_device(requester.user.to_string())
-        if dehydrated_device:
+        requester = await self.auth.get_user_by_req(request)
+        dehydrated_device = await self.device_handler.get_dehydrated_device(
+            requester.user.to_string()
+        )
+        if dehydrated_device is not None:
+            (device_id, device_data) = dehydrated_device
             result = {"device_id": device_id, "device_data": dehydrated_device}
             return (200, result)
         else:
-            raise errors.NotFoundError()
+            raise errors.NotFoundError("No dehydrated device available")
 
     async def on_PUT(self, request: SynapseRequest):
         submission = parse_json_object_from_request(request)
         requester = await self.auth.get_user_by_req(request)
 
+        if "device_data" not in submission:
+            raise errors.SynapseError(
+                400, "device_data missing", errcode=errors.Codes.MISSING_PARAM,
+            )
+        elif isinstance(submission["device_id"], dict):
```

Comment:
I think this is missing a `not`:

```suggestion
        elif not isinstance(submission["device_id"], dict):
```